### PR TITLE
draft: dpdk: Introduce support for DPDK secondary mode for workers runmode

### PR DESCRIPTION
**Pull request by lukashino**
Implement secondary mode operation in Suricata. Receive
packets over DPDK rings. To each worker assign a separate ring.
Provide configuration options specific for the secondary mode
along with possibility to specify arbitrary ring name.
Change threading from threads to DPDK lcores.
Add relevant documentation.

Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata.io/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/issues/4910) ticket:

Changelog from #7092:
- separate the change from pthreads to DPDK lcores to a single commit and address the reasons
- get DPDK lcore list from threading.cpu-affinity settings

### CODE REVIEW

File: src/runmode-dpdk.c [hunk 1]
```diff
@@ -247,6 +259,113 @@ static void ArgumentsAddOptionAndArgument(struct Arguments *args, const char *op
...
+ * Used for EAL initialization together with "-l" parameter (format -l 2,4,6,8).
+ * @return dynamically allocated string
+ */
+static char *ConfigLcoreArgValGet(void)
+{
+    SCEnter();
+    int ret;
+    uint16_t lcore_arr[sizeof(cpu_set_t)];
+    uint16_t arr_len = 0;
+    lcore_arr[arr_len++] = ConfigLcoreMainGet();
```
Comment (victorjulien) [inline src/runmode-dpdk.c:310]: seems a bit hacky... can we just start with `lcore_arr[0] = ConfigLcoreMainGet();`
followed by `uint16_t arr_len = 1 + ConfigLcoreWorkersSet(&lcore_arr[arr_len], sizeof(cpu_set_t) - 1);` or similar?

File: src/runmode-dpdk.c [hunk 2]
```diff
@@ -247,6 +259,113 @@ static void ArgumentsAddOptionAndArgument(struct Arguments *args, const char *op
...
+    uint16_t max_num = lcore_arr[arr_len - 1];
+
+    // arr_len - how many CPUs are set
+    // CountDigits(max_num) + 1 - the highest number of digits with comma separator
+    // +1 - terminating character
+    uint16_t lcore_arg_size = arr_len * (CountDigits(max_num) + 1) + 1;
+    char *lcore_arg = AllocArgument(lcore_arg_size);
+    uint16_t lcore_arg_len = 0;
+
+    for (int i = 0; i < arr_len; i++) {
```
Comment (victorjulien) [inline src/runmode-dpdk.c:321]: `i` should be `uint16_t` here

File: src/runmode-dpdk.c [hunk 3]
```diff
@@ -247,6 +259,113 @@ static void ArgumentsAddOptionAndArgument(struct Arguments *args, const char *op
...
+    SCReturnCharPtr(lcore_arg);
+}
+
+static void ArgumentsLcoreValidate(void)
+{
+    if (threading_set_cpu_affinity == FALSE)
+        FatalError(SC_ERR_DPDK_CONF, "CPU affinity needs to be set");
+
+    ThreadsAffinityType *mngmt_taf = GetAffinityTypeFromName("management-cpu-set");
+    ThreadsAffinityType *wrkr_taf = GetAffinityTypeFromName("worker-cpu-set");
```
Comment (victorjulien) [inline src/runmode-dpdk.c:345]: check for NULL on both, before unconditional deref later

File: src/runmode-dpdk.c [hunk 4]
```diff
@@ -612,6 +733,24 @@ static int ConfigSetCopyIfaceSettings(DPDKIfaceConfig *iconf, const char *iface,
     SCReturnInt(0);
 }
 
+static void ConfigSetOperationMode(DPDKIfaceConfig *iconf, const char *entry_str)
+{
+    enum rte_proc_type_t processType = rte_eal_process_type();
```
Comment (victorjulien) [inline src/runmode-dpdk.c:738]: nit: `process_type`

File: src/runmode-dpdk.c [hunk 5]
```diff
@@ -612,6 +733,24 @@ static int ConfigSetCopyIfaceSettings(DPDKIfaceConfig *iconf, const char *iface,
...
+
+    if (strcmp(entry_str, DPDK_CONFIG_OPERATION_MODE_ETHDEV) == 0 &&
+            processType == RTE_PROC_PRIMARY) {
+        iconf->op_mode = DPDK_ETHDEV_MODE;
+    } else if (strcmp(entry_str, DPDK_CONFIG_OPERATION_MODE_RING) == 0 &&
+               processType == RTE_PROC_SECONDARY) {
+        iconf->op_mode = DPDK_RING_MODE;
+    } else {
+        SCLogWarning(SC_WARN_DPDK_CONF,
```
Comment (victorjulien) [inline src/runmode-dpdk.c:747]: whats the reason to make this a warning instead of a `FatalError`?

File: src/runmode-dpdk.c [hunk 6]
```diff
@@ -1102,6 +1248,86 @@ static int DeviceConfigureIPS(DPDKIfaceConfig *iconf)
...
 }
 
+static void DeviceRingNameInit(char *name_out, uint8_t name_len, const char *format, uint16_t q_num)
+{
+    char name[RTE_RING_NAMESIZE];
+    char q_num_str[5];
+    char *q_pos;
+
+    if (q_num > 9999)
+        FatalError(SC_ERR_DPDK_CONF, "Number of queues must not exceed 10000 queues per interface");
```
Comment (victorjulien) [inline src/runmode-dpdk.c:1258]: 9999 or 10000?

File: src/runmode-dpdk.c [hunk 7]
```diff
@@ -1102,6 +1248,86 @@ static int DeviceConfigureIPS(DPDKIfaceConfig *iconf)
...
+    q_pos = strstr(format, DPDK_CONFIG_DEFAULT_QUEUE_NUM_SPECIFIER);
+    snprintf(name, MIN((long)sizeof(name), q_pos - format + 1), "%s", format);
```
Comment (victorjulien) [inline src/runmode-dpdk.c:1260]: can we get rid of such trickery when dealing with buffer sizes?

File: src/runmode-dpdk.c [hunk 8]
```diff
@@ -1102,6 +1248,86 @@ static int DeviceConfigureIPS(DPDKIfaceConfig *iconf)
...
+    iconf->rx_rings = SCCalloc(iconf->threads, sizeof(struct rte_ring *));
+    if (iconf->rx_rings == NULL) {
+        SCLogError(SC_ERR_DPDK_INIT, "Failed to calloc rx rings");
+        SCReturnInt(-ENOMEM);
+    }
+
+    iconf->tx_rings = SCCalloc(iconf->threads, sizeof(struct rte_ring *));
+    if (iconf->tx_rings == NULL) {
+        SCLogError(SC_ERR_DPDK_INIT, "Failed to calloc tx rings");
+        SCReturnInt(-ENOMEM);
```
Comment (victorjulien) [inline src/runmode-dpdk.c:1296]: are we leaking `rx_rings` here or is that handled somewhere else? (in this case, add comment)

File: src/runmode-dpdk.c [hunk 9]
```diff
@@ -1102,6 +1248,86 @@ static int DeviceConfigureIPS(DPDKIfaceConfig *iconf)
...
+    }
+
+    if (!DeviceRingNameIsValid(iconf->iface))
```
Comment (victorjulien) [inline src/runmode-dpdk.c:1299]: in general I prefer checks like this to be done before we make allocations we then have to unwind

File: src/runmode-dpdk.c [hunk 10]
```diff
@@ -1102,6 +1248,86 @@ static int DeviceConfigureIPS(DPDKIfaceConfig *iconf)
...
+        iconf->rx_rings[i] = rte_ring_lookup(name);
+        if (iconf->rx_rings[i] == NULL) {
+            SCLogError(SC_ERR_DPDK_INIT, "rte_ring_lookup(): cannot get rx ring '%s'", name);
+            SCReturnInt(-ENOENT);
+        }
+
+        if (iconf->copy_mode == DPDK_COPY_MODE_NONE) {
+            iconf->tx_rings[i] = NULL;
+        } else {
+            if (!DeviceRingNameIsValid(iconf->out_iface))
```
Comment (victorjulien) [inline src/runmode-dpdk.c:1315]: can be checked outside of the loop?

File: src/runmode-dpdk.c [hunk 11]
```diff
@@ -1269,13 +1495,27 @@ static void *ParseDpdkConfigAndConfigureDevice(const char *iface)
...
+            FatalError(SC_ERR_DPDK_CONF, "Interface \"%s\" unable to obtain port ID: %s",
+                    iconf->iface, rte_strerror(-retval));
+        }
 
-        FatalError(SC_ERR_DPDK_CONF, "Device %s fails to configure", iface);
+        if (DeviceConfigure(iconf) != 0) {
+            iconf->DerefFunc(iconf);
+            retval = rte_eal_cleanup();
+            if (retval != 0)
+                FatalError(SC_ERR_DPDK_EAL_INIT, "EAL cleanup failed: %s", strerror(-retval));
```
Comment (victorjulien) [inline src/runmode-dpdk.c:1515]: I guess the error here should say something like "device %s failed to configure, additionally couldn't properly shutdown eal"?

File: src/source-dpdk.h [hunk 12]
```diff
@@ -31,6 +31,11 @@ typedef enum { DPDK_COPY_MODE_NONE, DPDK_COPY_MODE_TAP, DPDK_COPY_MODE_IPS } Dpd
 
 #define DPDK_BURST_TX_WAIT_US 1
 
+typedef enum {
+    DPDK_ETHDEV_MODE,
```
Comment (victorjulien) [inline src/source-dpdk.h:35]: comment: "primary mode"

File: src/source-dpdk.h [hunk 13]
```diff
@@ -31,6 +31,11 @@ typedef enum { DPDK_COPY_MODE_NONE, DPDK_COPY_MODE_TAP, DPDK_COPY_MODE_IPS } Dpd
...
+    DPDK_RING_MODE,
```
Comment (victorjulien) [inline src/source-dpdk.h:36]: comment: "secondary mode"

File: src/source-dpdk.h [hunk 14]
```diff
@@ -43,8 +48,13 @@ typedef struct DPDKIfaceConfig_ {
     char iface[RTE_ETH_NAME_MAX_LEN];
     uint16_t port_id;
     uint16_t socket_id;
+    DpdkOperationMode op_mode;
     /* number of threads - zero means all available */
     int threads;
+    /* Ring mode settings */
+    struct rte_ring **rx_rings;
+    struct rte_ring **tx_rings;
```
Comment (victorjulien) [inline src/source-dpdk.h:56]: comment should explain why these are here, and how they relate to `tx_ring`

File: src/tm-threads.c [hunk 15]
```diff
@@ -1719,31 +1779,46 @@ void TmThreadClearThreadsFamily(int family)
...
 {
+    int rc;
     pthread_attr_t attr;
     if (tv->tm_func == NULL) {
         FatalError(SC_ERR_TM_THREADS_ERROR, "No thread function set");
     }
 
-    /* Initialize and set thread detached attribute */
-    pthread_attr_init(&attr);
+#ifdef HAVE_DPDK
```
Comment (victorjulien) [inline src/tm-threads.c:1788]: could we split this out to have 2 util funcs `TmThreadSpawnCreate` and `TmThreadSpawnCreateDPDK` or something similar?

File: src/util-dpdk.c [hunk 16]
```diff
@@ -18,16 +18,32 @@
...
+ * \author Lukas Sismis <sismis@cesnet.cz>
  */
 
+#ifndef UTIL_DPDK_C
+#define UTIL_DPDK_C
+
 #include "suricata.h"
 #include "util-dpdk.h"
 
+uint8_t CountDigits(uint16_t n)
```
Comment (victorjulien) [inline src/util-dpdk.c:30]: I'm dumb. I really don't understand the what/why/how here

File: src/util-runmodes.c [hunk 17]
```diff
@@ -253,8 +253,20 @@ static int RunModeSetLiveCaptureWorkersForDevice(ConfigIfaceThreadsCountFunc Mod
                               const char *live_dev, void *aconf,
                               unsigned char single_mode)
 {
+#ifdef HAVE_DPDK
+    // equal to -1 to get the first lcore id
+    int16_t dpdk_last_spawned_lcore = -1;
```
Comment (victorjulien) [inline src/util-runmodes.c:258]: I think I've seen `lcore_id` as uint32_t, uint16_t and int16_t. Can we settle on one type?

**Review by victorjulien — CHANGES_REQUESTED** [review]

File: src/source-dpdk.h (hunk 12 above)
Comment (lukashino) [reply src/source-dpdk.h:35]: comment added

File: src/source-dpdk.h (hunk 13 above)
Comment (lukashino) [reply src/source-dpdk.h:36]: comment added

File: src/tm-threads.c (hunk 15 above)
Comment (lukashino) [reply src/tm-threads.c:1788]: Did try to remove those ifdef things as much as possible here, or at least moved them to a different function. By moving it to another function I was able to mitigate the issue of wrapped `if` in ifdefs.

**Review by lukashino — COMMENTED** [review]
Reviewed and implemented your comments, I am opening a new PR soon.

**Comment by suricata-qa** [comment]
Warning: no commits in this PR have specified the following ticket(s):
- 4910 - https://redmine.openinfosecfoundation.org/issues/4910

Please update the commit(s) and submit a new PR.

**Comment by catenacyber** [comment]
Replaced by #7283, right ?

**Pull request closed without merging**